### ISSUE TITLE
conflating keys and nodes

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaCollections/AbstractTrees.jl#98]
I have just come to fully understand a problem that I was struggling with a bit when I wrote my 0.4 PR: current master on this package still conflates "keys" with "nodes".

To see what I mean by this, consider a `Dict`.  Currently we define `children(dict::AbstractDict) = pairs(dict)` and some methods for `Pair` so that `Pair`s are by themselves nodes each with a single child (i.e. they are tree-equivalent to `[p.second]`.  The problem with this is that `p.first` is supposed to play the role of an index, not a node.  This matters because, for example, `getdescendant` (used to be called `childindex` but I changed it after a helpful suggestion by @jlumpe) cannot work with dictionaries, even though it has every reason to.

It's mostly clear to me now that the "right" thing to do is to deprecate the old behavior so that e.g. `children(dict::AbstractDict) = values(dict)` and delete all of the `pair` methods.  However, this is making me a little nervous.  My PR doesn't break very much in part because the tree structure of everything in `Base` is preserved.  Making alterations to this tree structure is much more likely to be breaking.  One place this would be manifest within AbstractTrees itself is `print_tree`, which would need to be altered.

In summary, the benefit of making this change would be consistent indexing behavior, with the cost of making what is probably a more significant change than any that were in my original PR.  On the other hand, it probably won't break all that many packages.

Opinions on this?

**Comment by ExpandingMan** [comment]
Small update: I just remembered that `getindex` works on `pairs(dict)` just like it does on `dict`.  This has the implication for `Dict` that `getdescendant` works even on current master.

That probably means we shouldn't change anything: the most obvious benefit of making this change is already present, so it's hard to see what we'd be getting in return for a breaking change like this other than philosophical consistency.

**Comment by jlumpe** [comment]
I had filed a PR (#86) to resolve this problem. It defines a `DictChildren` type which behaves identically to the wrapped `AbstractDict` with regard to `getindex()`, `keys()`, `values()`, and `pairs()`, but acts as a collection of the dict's values instead of its pairs. It's fairly out of date at this point but at the time I wrote it it worked with printing. It would need a bit of work to adapt it to the latest changes but I think it's the best solution to this problem.

**Comment by ExpandingMan** [comment]
I think we should avoid complicating the interface.

Since my realization that `pairs(dict)` can be indexed it seems very unlikely that we will change this.  Certainly we would not get any benefits out of such a breaking change in the current PR.

Unless someone can come up with a more compelling reason why we should "fix" it, probably best to drop this matter.

**Comment by ExpandingMan** [comment]
The other problem is that `values` does not return an object which has `getindex` methods for the keys.  I think a lot of this goes back to it probably having been a mistake for `Base` to make dictionaries iterables of pairs rather than more array-like objects which are indexed by their keys rather than integers.  This is the reason that [Dictionaries.jl](https://github.com/andyferris/Dictionaries.jl) exists.

I suppose I can check how much a change would really be breaking, I'd certainly welcome help with that.  Off the top of my head, the only "correct" solution I can think of would be something like your `DictChildren` suggestion where we return an iterator over the values but wrap it in something which has `getindex` methods.

**Comment by oscardssmith** [comment]
none

**Comment by ExpandingMan** [comment]
Another perspective is that the problem is with `Base`, not AbstractTrees, and we are merely adhering to the somewhat-regrettable interface that it established.  Users who want better behavior can use Dictionaries.jl.  Certainly if `Base` worked the way Dictionaries.jl did, there'd be no question in my mind that we should fix this regardless of how breaking it is (it probably would not have been done this way in the first place).

🤷 I don't know, I think we have to at least try to figure out what this is likely to break.  Again, if anyone sees anything, please post here.
